pkg/plugin/metrics_functions.go

```diff
@@ -0,0 +1,1157 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/oracle/oci-go-sdk/v65/common"
+	"github.com/oracle/oci-go-sdk/v65/identity"
+	"github.com/oracle/oci-go-sdk/v65/monitoring"
+	"github.com/oracle/oci-grafana-metrics/pkg/plugin/constants"
+	"github.com/oracle/oci-grafana-metrics/pkg/plugin/models"
+	"github.com/pkg/errors"
+)
+
+type metricDataBank struct {
+	dataPoints     []monitoring.MetricData
+	resourceLabels map[string]map[string]string
+}
+
+// TestConnectivity Check the OCI data source test request in Grafana's Datasource configuration UI.
+func (o *OCIDatasource) TestConnectivity(ctx context.Context) error {
+	backend.Logger.Debug("client", "TestConnectivity", "testing the OCI connectivity")
+
+	var reg common.Region
+	var testResult bool
+
+	tenv := o.settings.Environment
+	tmode := o.settings.TenancyMode
+
+	for key, _ := range o.tenancyAccess {
+		testResult = false
+
+		if tmode == "multitenancy" && tenv == "OCI Instance" {
+			return errors.New("Multitenancy mode using instance principals is not implemented yet.")
+		}
+		tenancyocid, tenancyErr := o.tenancyAccess[key].config.TenancyOCID()
+		if tenancyErr != nil {
+			return errors.Wrap(tenancyErr, "error fetching TenancyOCID")
+		}
+
+		regio, regErr := o.tenancyAccess[key].config.Region()
+		if regErr != nil {
+			return errors.Wrap(regErr, "error fetching Region")
+		}
+		reg = common.StringToRegion(regio)
+		o.tenancyAccess[key].monitoringClient.SetRegion(string(reg))
+
+		// Test Tenancy OCID first
+		backend.Logger.Debug(key, "Testing Tenancy OCID", tenancyocid)
+		listMetrics := monitoring.ListMetricsRequest{
+			CompartmentId: &tenancyocid,
+		}
+
+		var status int
+		res, err := o.tenancyAccess[key].monitoringClient.ListMetrics(ctx, listMetrics)
+		if err != nil {
+			backend.Logger.Debug(key, "SKIPPED", err)
+		} else {
+			status = res.RawResponse.StatusCode
+		}
+		if status >= 200 && status < 300 {
+			backend.Logger.Debug(key, "OK", status)
+		} else {
+			backend.Logger.Debug(key, "SKIPPED", fmt.Sprintf("listMetrics on Tenancy %s did not work, testing compartments", tenancyocid))
+			comparts := o.GetCompartments(ctx, tenancyocid)
```

Comment:
Do we need a check to see if GetCompartments() return nil ?

Reply:
Here, and in general when there are function of the type o.<something> the error handling is performed, in a more granular way, within the called function

Reply:
however I added the error handling as you suggesdted :)